crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+                self.operator = Operation::ult
+            }
+            Operation::uge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ule
+            }
+            Operation::sgt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::slt
+            }
+            Operation::sge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::sle
+            }
+            Operation::gt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lt
+            }
+            Operation::gte => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lte
+            }
+            //TODO replace a<b with a<=b+1, but beware of edge cases!
+            Operation::eq_gate => {
+                if self.rhs == self.lhs {
+                    self.rhs = self.idx;
+                    self.is_deleted = true;
+                }
+            }
+            _ => (),
+        }
+        if is_commutative(self.operator) && self.rhs < self.lhs {
+            std::mem::swap(&mut self.rhs, &mut self.lhs);
+        }
+    }
+}
+
+//adpated from LLVM IR
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Operation {
+    add,      //(+)
+    safe_add, //(+) safe addtion
+    sub,      //(-)
+    safe_sub, //(-) safe substraction
+    mul,      //(*)
+    safe_mul, //(*) safe multiplication
+    udiv,     //(/) unsigned division
+    sdiv,     //(/) signed division
+    urem,     //(%) modulo; remainder of unsigned division
+    srem,     //(%) remainder of signed division
+    fmod,     //(%) remainder of the floating point division
+    fneg,     //(-) negation of a float
+    fdiv,     //(/) floating point division
+    div,      //(/) field division
+    eq,       //(==) equal
+    ne,       //(!=) not equal
+    ugt,      //(>) unsigned greater than
+    uge,      //(>=) unsigned greater or equal
+    ult,      //(<) unsigned less than
+    ule,      //(<=) unsigned less or equal
+    sgt,      //(>) signed greater than
+    sge,      //(>=) signed greater or equal
+    slt,      //(<) signed less than
+    sle,      //(<=) signed less or equal
+    lt,       //(<) field less
+    gt,       //(>) field greater
+    lte,      //(<=) field less or equal
+    gte,      //(<=) field greater or equal
+    and,
+    not,
+    or,
+    xor,
+    cast,  //convert type
+    ass,   //assignement
+    trunc, //truncate
+
+    //control flow
+    jne, //jump on not equal
+    jeq, //jump on equal
+    jmp, //unconditional jump
+    phi,
+    // todo: call, br,..
+    nop, // no op
+
+    eq_gate, //write a gate enforcing equality of the two sides (to support the constrain statement)
+             //memory todo: load, store, getelementptr?
+}
```

Comment:
Lets open up a tracking issue for memory operations and remove this commment

Reply:
Done: https://github.com/noir-lang/noir/issues/157